Add Support for Different Transaction Types

dzou opened the pull request:
Adds an additional overloaded method `beginTransaction(transactionType)` to `SpannerConnection` in order to support multiple transaction types: read-write, read-only, and partitioned DML.

Notes:
- Commit Transaction is not a valid operation for read-only and partitioned DML transactions. Thus, commit will be a no-op if called during one of these transaction types.
- For partitioned dml statements, exact row count updated is not returned but a lower bound is returned instead. Thus this number is used to populate SpannerResult.

In reference to: #40.

ChengyuanZhao commented on src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java:
```diff
@@ -89,15 +88,14 @@ public GrpcClient(GoogleCredentials credentials) {
-                  TransactionOptions
-                      .newBuilder()
-                      .setReadWrite(ReadWrite.getDefaultInstance()))
+              .setOptions(transactionOptions)
```
In addition to setting the transaction options here, don't you also need to actually swap out the call for from executeStreamingSql to partitionQuery for example?

dzou replied:
Hmm am I mistaken about the Partitioned DML - but that is executed through executeStreamingSQL and not partitionQuery right?

see: https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#partitionqueryrequest
> This must not contain DML commands, such as INSERT, UPDATE, or DELETE. Use ExecuteStreamingSql with a PartitionedDml transaction for large, partition-friendly DML operations.

meltsufin commented on src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java:
```diff
@@ -31,6 +33,11 @@
  */
 public class SpannerConnection implements Connection {
 
+  private static final TransactionOptions READ_WRITE_TRANSACTION =
```
Does it make sense to just create our own `TransactionType` enum and not expose this gRPC level object to the clients?

dzou replied:
Sounds good. For this one I realized that the READ-ONLY `transactionType` actually has a number of different options that the user can set: https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.TransactionOptions.ReadOnly

@elefeint, @meltsufin What are your thoughts on exposing these settings for read-only? Should we make this into a TransactionType class then with builder pattern?

meltsufin replied:
In that case I would just expose `TransactionOptions` like you have it now. Ignore my enum suggestion.

dzou replied:
Ok, leaving as is for now.

meltsufin commented on src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java:
```diff
@@ -53,35 +60,42 @@ public SpannerConnection(Client client, Session session) {
-        .doOnNext(
-            transaction -> this.transactionContext = SpannerTransactionContext.from(transaction))
+  public Mono<Void> beginTransaction() {
+    return this.beginTransaction(READ_WRITE_TRANSACTION);
```
Should the default come from the `ConnectionOptions` or something?

elefeint replied:
This could be in scope for #119.

dzou replied:
Yeah for this PR I will only focus on adding a hook for setting transaction through `beginTransaction`; the connection options propagation will be deferred to a followup PR.

elefeint commented on src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java:
```diff
@@ -53,35 +60,42 @@ public SpannerConnection(Client client, Session session) {
+
+      if (this.transactionContext == null) {
+        this.logger.debug("commitTransaction() is a no-op; called with no transaction active.");
+      } else if (this.transactionContext.isReadWrite()) {
```
Should this be negated? The debug message says we are *not* in read-write here.

dzou replied:
Nice catch, done.

ChengyuanZhao commented on src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java:
```diff
@@ -80,6 +83,21 @@
+          .setReadWrite(ReadWrite.getDefaultInstance())
+          .build();
+
+  private static final TransactionOptions PARTITIONED_DML =
```
There might be some messiness when merging w/ Dmitry's batch code, because there we check for the query string's starting word and use the batch execute method for all DML. That might clash with the partitioned DML request type you have here.

dzou replied:
Yeah I ran into that just now. Seems like one cannot use the BatchDDL method to run partitioned DML statements.

dzou commented:
> Looks good; `SpannerConnectionTest` needs new tests for corresponding functionality changes, and there may be a negation needed in an edge case.

Done. Added a test to `SpannerConnectionTest` to verify that the transaction options are propagated and also commit is a no-op for non-readWrite transactions. Also added an integration test with multiple transaction types.
